add check_mode option for tasks

robinro opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

devel, 2.2
##### SUMMARY

Introduce an option `check_mode` that can be added to tasks. Forces or prevents the module to run under check_mode.

Replaces the `always_run` option (now: `check_mode: no`) but extends it by allowing to run under check mode even if check mode is not specified on the command line (`check_mode: yes`).

Fixes #12498. See that issue and links therein also for discussion why this is nice.

Includes example testcases for the template module. One of the benefits of including per-task check_mode is that testing, both of ansible itself and external projects using ansible, becomes easier.

Commit:
add check_mode option for tasks

includes example testcases for the template module

robinro commented:
btw: I was quite impressed when I realized that this is actually a 3-line patch. Ansible is really flexible!

bcoca commented:
the patch does not handle correctly when check mode is specified in the command line but the task has it set to false.

Commit:
extend check_mode option

* replace always_run, see also proposal rename_always_run
* rename always_run where used and add deprecation warning
* add some documentation

robinro commented:
The case where check mode is given on the command line but we still want to run the task was so far covered by `always_run`. I extended the implementation to replace `always_run` by `check_mode: no` and have `check_mode: yes` to enforce check mode even if it is not given on the command line.

This implements/extends the _rename_always_run_ proposal by @resmo: https://github.com/ansible/proposals/blob/master/rename_always_run.md.

bcoca commented on lib/ansible/playbook/play_context.py:
```diff
@@ -434,8 +434,12 @@ def set_task_and_variable_override(self, task, variables, templar):
         # set become defaults if not previouslly set
         task.set_become_defaults(new_info.become, new_info.become_method, new_info.become_user)
 
+        if task.check_mode is not None:
```
~~you still don't deal with `check_mode: False` and~~ now always_run will override check_mode setting, I would reverse it in view of deprecating always_run

Commit:
have check_mode overwrite always_run

robinro commented:
@bcoca makes sense, I switched the two to have `check_mode` overwrite `always_run`

bcoca commented:
nice, now we just need enough committers to agree.

Commit:
use unique template name to prevent conflicts

test_check_mode was right before, but failed due to using the same filename as other roles

Commit:
still mention always_run in the docs

Commit:
set deprecation of always_run to version 2.4

abadger commented on docsite/rst/playbooks_checkmode.rst:
```diff
@@ -20,27 +20,27 @@ Example::
+
+If the playbook is run with checkmode on the command line, you can force tasks to still run, by specifying `check_mode: no`.
+On the other hand you can setup a test section which, independetly of the command line arguments, always runs a tasks in check mode with `check_mode: yes`.
+note:: Prior to version 2.2 only the ability to turn this on per-task existed.  The notation for that was `always_run: yes`.
```
For proper rst syntax, this should be:

```
.. note:: Prior to[...]
```

abadger commented:
+1 from me.

Commit:
fix rst style

robinro commented:
Based on the discussion in the core meeting I updated the docs (still mention always_run) and changed the version for deprecating always_run to 2.4.

Thanks for the review!

alxgu commented:
+1

robinro commented:
@abadger @bcoca Can this be merged? In the core-meeting discussion 3 weeks ago you approved it: https://meetbot.fedoraproject.org/ansible-meeting/2016-06-09/ansible_core_team_public_meeting.2016-06-09-15.11.log.html

The shippable failure is unrelated afaik.

resmo commented on lib/ansible/playbook/play_context.py:
```diff
@@ -434,10 +434,15 @@ def set_task_and_variable_override(self, task, variables, templar):
 
-        # have always_run override check mode
         if task.always_run:
+            display.deprecated("always_run is deprecated. Use check_mode = no instead.", version="2.4", removed=False)
```
I would change the message to just: `always_run is deprecated. Use check_mode instead`

robinro replied:
In 2.1 one used `always_run: yes` and with the short message one might try `check_mode: yes`, which is the opposite.

amenonsen commented:
As documented and implemented, I find this feature extremely confusing and unintuitive. I understand the motivation, and think the underlying feature is desirable (in fact, I submitted a PR for it years ago), but I don't think it should be merged without revisions.

robinro commented:
@amenonsen can you be more specific and suggest improvements?
I thought about this change a lot and think that check_mode: yes/no is the most straight-forward way to call it and also does not conflict with the previous implementation.

bcoca commented:
@amenonsen can you expand on your comments? otherwise i'll be merging this soon. 

I do think we need to have very clear and explicit docs on how this is different than 'always_run' and how it makes it obsolete.

Commit:
expand documentation on per-task check mode

mattclay commented:
The expanded documentation on this helps. I think the only thing that would make it more clear is if the options for check_mode where "always" or "never", instead of a boolean. However, I'm willing to accept this as is.

robinro commented:
Thanks for merging @bcoca

dagwieers commented:
According to the documentation:

> There are two options:
> - Force a task to run in check mode, even when the playbook is called without --check. This is called check_mode: yes.
> - Force a task to run in normal mode and make changes to the system, even when the playbook is called with --check. This is called check_mode: no.

I understand where we are coming from, and see how you now have more possibilities to get something run despite being in check-mode. But having to add `check_mode: no` to a task that needs to run in check-mode as well as normal run-mode (which is to most common use-case) is so counter-intuitive that it becomes a big usability problem IMO.

Compare this to: `always_run: yes` and you can see how this is very clear to what it means, whereas `check_mode: no` does the opposite of what you would assume it would do. The boolean does not properly indicate what it is switching on/off.

This is far from the idea that a playbook is human-readable and easy to understand. And I would propose to revert the deprecation of `always_run` in v2.0.1 **and** replace the check_mode option with something that is far more readable. (before people start using this...)

robinro commented:
@dagwieers Can you propose a better notation? Or is your proposal to go back to `always_run` and not unify the `check_mode: yes/no` options in one variable?

I like the `check_mode: yes/no` notation. The tasks `run` always anyway, but sometimes they run in `check_mode` and sometimes they don't. There is a flag on the command line to enable/disable checkmode for all tasks and if you set `check_mode: yes/no` on a task it overwrites the global flag.

If you are a `always_run` user it might be countintuitive, but if you start unbiased I feel this is the right setup.

dagwieers commented:
The problem is we are using the term _check_mode_ for two distinct things. One such thing is the mode we are running in and the other thing is how we want the task to behave when being run. And that is causing the confusion.

And as such we have 4 different cases:
- Execute when in run-mode, run task as if in run-mode (normal behaviour)
- Execute when in check-mode, run task as if in check-mode (normal behaviour when --check)
- Execute when in run-mode, run task as if in check-mode (check_mode: yes)
- Execute when in check-mode, run task as in run-mode (check_mode: no)

The option `always_run` was straightforward because it only applied to the mode we were running in. And simply meant run it in both run-mode as well as in check-mode. But the `check_mode` option is diffusing all this by adding the potential to modify _how_ a task is being run, not just _when_ it ought to run.

So if we would want to allow common combinations, an option could be

``` yaml
    run_mode: [ check, normal ]
    behave: check/normal
```

And the default behaviour is that a task behaves as the run-mode it runs in. Another way to write it out could be specify the behaviour for each running mode (even more options):

``` yaml
    check_mode: check/run
    run_mode: run/check
```

and while that is very explicit and even gives more freedom (you could make it do the exact opposite), I would prefer we introduce new terminology for the task-behavior (operate, behave, act, ...). Why ? Because run-mode/check-mode/diff-mode are already known/loaded terms. Having new terminology will make it easier to communicate with one another, especially for giving support.

Another issue I have with the current design (but also with _always_run_) is the fact that it limits to the two run-modes we have today, but we may have other run-modes in the future (in fact, diff-mode + check-mode is already close to a report-mode I envisioned once) and the current design. So I would prefer a solution that doesn't forces us in the existing model. (E.g. some users may want to always enable diff-mode for specific tasks.)

So unfortunately I think we jumped the gun on this one.
